File: core/src/net/namekdev/memcop/view/CodeCursor.java
package net.namekdev.memcop.view;

import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.graphics.g2d.Batch;
import com.badlogic.gdx.scenes.scene2d.Actor;
import com.badlogic.gdx.scenes.scene2d.Touchable;
import com.kotcrab.vis.ui.widget.HighlightTextArea;
import net.namekdev.memcop.Assets;
import net.namekdev.memcop.domain.Assembly;
import net.namekdev.memcop.domain.Transputer;

public class CodeCursor extends Actor {
    private static final Color COLOR_CURSOR = Color.valueOf("790000");

    private HighlightTextArea codeInput;
    private Transputer transputer;

    public CodeCursor(HighlightTextArea codeInput, Transputer transputer) {
        this.codeInput = codeInput;
        this.transputer = transputer;

        setTouchable(Touchable.disabled);
    }

    @Override
    public void draw(Batch batch, float parentAlpha) {
        super.draw(batch, parentAlpha);

        final Assembly.Instruction instr = transputer.getNextInstruction();
        if (!transputer.isDebuggingActive())
            return;

        float lineHeight = codeInput.getStyle().font.getLineHeight();
        float y, h;

        if (instr != null) {
            y = codeInput.getHeight() - instr.lineNumber * lineHeight + 4;
            h = lineHeight - 2;
        }
        else {
            // render bottom line when whole program is finished
            y = codeInput.getHeight() - transputer.lastInstruction.lineNumber * lineHeight + 4;
            h = 4;
        }

        batch.setColor(COLOR_CURSOR);
        batch.draw(Assets.white, 0, y, codeInput.getWidth(), h);
    }
}

File: core/src/net/namekdev/memcop/view/CodeCursor.kt
package net.namekdev.memcop.view

import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.graphics.g2d.Batch
import com.badlogic.gdx.scenes.scene2d.Actor
import com.badlogic.gdx.scenes.scene2d.Touchable
import com.kotcrab.vis.ui.widget.HighlightTextArea
import net.namekdev.memcop.Assets
import net.namekdev.memcop.domain.Assembly
import net.namekdev.memcop.domain.Transputer

class CodeCursor(private val codeInput: HighlightTextArea, private val transputer: Transputer) : Actor() {

    init {
        touchable = Touchable.disabled
    }

    override fun draw(batch: Batch, parentAlpha: Float) {
        super.draw(batch, parentAlpha)

        val instr = transputer.nextInstruction
        if (!transputer.isDebuggingActive)
            return

        val s = Render.scale
        val lineHeight = codeInput.style.font.lineHeight
        val y: Float
        val h: Float

        if (instr != null) {
            y = codeInput.height - instr.lineNumber * lineHeight + 4 * s
            h = lineHeight - 2 * s
        }
        else {
            // render bottom line when whole program is finished
            y = codeInput.height - transputer.lastInstruction!!.lineNumber * lineHeight + 4 * s
            h = 4f * s
        }

        batch.color = COLOR_CURSOR
        batch.draw(Assets.white, 0f, y, codeInput.width, h)
    }

    companion object {
        private val COLOR_CURSOR = Color.valueOf("790000")
    }
}
